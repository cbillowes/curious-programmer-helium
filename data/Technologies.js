module.exports = [
  {
    name: "Clojure",
    current: true,
    start: new Date(2018, 7, 1)
  },
  {
    name: "Datomic",
    current: true,
    start: new Date(2018, 7, 1)
  },
  {
    name: "C#",
    xp: 10
  },
  {
    name: "React",
    current: true,
    start: new Date(2018, 6, 26)
  },
  {
    name: "Git",
    current: true,
    start: new Date(2014, 4, 1)
  },
  {
    name: "Gatsby",
    current: true,
    start: new Date(2018, 6, 26)
  },
  {
    name: "T-SQL",
    xp: 11
  },
  {
    name: "JavaScript",
    current: true,
    start: new Date(2006, 1, 1)
  },
  {
    name: "jQuery",
    current: true,
    start: new Date(2009, 1, 1)
  },
  {
    name: "VB.NET",
    xp: 3
  },
  {
    name: "Jenkins",
    xp: 5
  },
  {
    name: "MVC"
  },
  {
    name: "Grunt"
  },
  {
    name: "Solr",
    current: true,
    start: new Date(2018, 12, 1)
  },
  {
    name: "WebAPI"
  },
  {
    name: "Handlebars"
  },
  {
    name: "LINQ"
  },
  {
    name: "Reagent",
    current: true,
    start: new Date(2018, 7, 1)
  },
  {
    name: "Web Forms"
  },
  {
    name: "Entity Framework"
  },
  {
    name: "Bootstrap"
  },
  {
    name: "Semantic UI",
    current: true,
    start: new Date(2018, 7, 1)
  },
  {
    name: "CSS/3",
    current: true,
    start: new Date(2004, 1, 1)
  },
  {
    name: "SASS/LESS",
    current: true
  },
  {
    name: "HTML/5",
    current: true,
    start: new Date(2004, 1, 1)
  },
  {
    name: "ClojureScript",
    current: true,
    start: new Date(2018, 7, 1)
  },
  {
    name: "Agile",
    xp: 5
  },
  {
    name: "Clickatel Integration",
    xp: 1
  },
  {
    name: "XMPie Integration",
    xp: 2
  },
  {
    name: "VCS Integration",
    xp: 1
  },
  {
    name: "Nedbank iVeri Lite Integration",
    xp: 3
  },
  {
    name: "Mashreq Payment Gateway Integration",
    xp: 1
  },
  {
    name: "Microsoft SQL Server",
    xp: 11
  }
]
