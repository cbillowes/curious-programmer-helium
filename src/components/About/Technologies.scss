@import "../../layout/layout.scss";

.technologies {
    margin-top: 1.5em;
    text-align: center;

    li {
        display: inline;
    }

    .tag {
        @include button;
        margin: 0.5em;
        
        .xp {
            font-size: 0.75em;
            color: lighten($background, 20%);

            &::before {
                content: " ";
            }
        }
    }

    .button {
        @include button;
        background-color: $foreground-accent;
        padding-bottom: 3px;
    }

    .elaborate {
        display: none;
    }

    .elaborated {
        display: block; 
    }
}