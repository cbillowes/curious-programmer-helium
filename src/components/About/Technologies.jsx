import React, { Component } from "react"
import technologies from "../../../data/Technologies"
import "./Technologies.scss"
const _ = require("lodash")

function getPrintedXP(years, months) {
  if (months >= 6) {
    years += 1
  }

  if (months > 0 && years === 0) return "( " + months + " mo )"
  if (years === 1) return "( 1 yr )"
  if (years > 1) return "( " + years + " yrs )"
  return "( < 1 mo )"
}

class Technologies extends Component {
  constructor(props) {
    super(props)

    this.state = {
      elaborated: false
    }
  }

  isElaborate = () => {
    return this.state.elaborated
  }

  toggle = () => {
    this.setState({ elaborated: !this.isElaborate() })
  }

  getXP = (existingXP, start) => {
    if (existingXP) {
      return getPrintedXP(existingXP, 0)
    }

    if (start) {
      const day = 1000 * 60 * 60 * 24
      let end = new Date()
      let diff = Math.floor(end.getTime() - start.getTime())
      let days = Math.floor(diff / day)
      let months = Math.ceil(days / 31)
      let years = Math.floor(months / 12)
      return getPrintedXP(years, months)
    }

    return ""
  }

  render() {
    const toptal = technologies.slice(0, 5)
    const hidden = technologies.slice(5, technologies.length)

    return (
      <div className="technologies">
        {toptal.map((item) => (
          <span
            key={`technologies-${_.kebabCase(item.name.toLowerCase())}`}
            className={`tag ${item.current ? "current" : ""}`}>
            {item.name}
            <span className="xp">{this.getXP(item.xp, item.start)}</span>
          </span>
        ))}
        <div className="button" onClick={this.toggle}>
          {this.isElaborate() ? "less" : "more"}
        </div>
        <div
          className={`elaborate ${this.isElaborate() ? "elaborated" : ""}`}
        >
          {hidden.map((item) => (
            <span
              key={`technologies-${_.kebabCase(item.name.toLowerCase())}`}
              className={`tag ${item.current ? "current" : ""}`}>
              {item.name}
              <span className="xp">{this.getXP(item.xp, item.start)}</span>
            </span>
          ))}
        </div>
      </div>
    )
  }
}

export default Technologies
