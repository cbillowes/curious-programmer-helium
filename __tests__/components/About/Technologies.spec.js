import React from "react"
import renderer from "react-test-renderer"
import { mount } from "enzyme"
import Technologies from "../../../src/components/About/Technologies"

describe("technologies component", () => {
  it("should render", () => {
    const component = <Technologies />
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("should render contents", () => {
    const component = mount(<Technologies />)
  })
})