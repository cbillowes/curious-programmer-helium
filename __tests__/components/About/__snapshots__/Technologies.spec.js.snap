// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`technologies component should render 1`] = `
<div
  className="technologies"
>
  <span
    className="tag current"
  >
    Clojure
    <span
      className="xp"
    >
      ( 1 yr )
    </span>
  </span>
  <span
    className="tag current"
  >
    Datomic
    <span
      className="xp"
    >
      ( 1 yr )
    </span>
  </span>
  <span
    className="tag "
  >
    C#
    <span
      className="xp"
    >
      ( 10 yrs )
    </span>
  </span>
  <span
    className="tag current"
  >
    React
    <span
      className="xp"
    >
      ( 1 yr )
    </span>
  </span>
  <span
    className="tag current"
  >
    Git
    <span
      className="xp"
    >
      ( 5 yrs )
    </span>
  </span>
  <div
    className="button"
    onClick={[Function]}
  >
    more
  </div>
  <div
    className="elaborate "
  >
    <span
      className="tag current"
    >
      Gatsby
      <span
        className="xp"
      >
        ( 1 yr )
      </span>
    </span>
    <span
      className="tag "
    >
      T-SQL
      <span
        className="xp"
      >
        ( 11 yrs )
      </span>
    </span>
    <span
      className="tag current"
    >
      JavaScript
      <span
        className="xp"
      >
        ( 13 yrs )
      </span>
    </span>
    <span
      className="tag current"
    >
      jQuery
      <span
        className="xp"
      >
        ( 10 yrs )
      </span>
    </span>
    <span
      className="tag "
    >
      VB.NET
      <span
        className="xp"
      >
        ( 3 yrs )
      </span>
    </span>
    <span
      className="tag "
    >
      Jenkins
      <span
        className="xp"
      >
        ( 5 yrs )
      </span>
    </span>
    <span
      className="tag "
    >
      MVC
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag "
    >
      Grunt
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag current"
    >
      Solr
      <span
        className="xp"
      >
        ( 1 mo )
      </span>
    </span>
    <span
      className="tag "
    >
      WebAPI
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag "
    >
      Handlebars
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag "
    >
      LINQ
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag current"
    >
      Reagent
      <span
        className="xp"
      >
        ( 1 yr )
      </span>
    </span>
    <span
      className="tag "
    >
      Web Forms
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag "
    >
      Entity Framework
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag "
    >
      Bootstrap
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag current"
    >
      Semantic UI
      <span
        className="xp"
      >
        ( 1 yr )
      </span>
    </span>
    <span
      className="tag current"
    >
      CSS/3
      <span
        className="xp"
      >
        ( 15 yrs )
      </span>
    </span>
    <span
      className="tag current"
    >
      SASS/LESS
      <span
        className="xp"
      >
        
      </span>
    </span>
    <span
      className="tag current"
    >
      HTML/5
      <span
        className="xp"
      >
        ( 15 yrs )
      </span>
    </span>
    <span
      className="tag current"
    >
      ClojureScript
      <span
        className="xp"
      >
        ( 1 yr )
      </span>
    </span>
    <span
      className="tag "
    >
      Agile
      <span
        className="xp"
      >
        ( 5 yrs )
      </span>
    </span>
    <span
      className="tag "
    >
      Clickatel Integration
      <span
        className="xp"
      >
        ( 1 yr )
      </span>
    </span>
    <span
      className="tag "
    >
      XMPie Integration
      <span
        className="xp"
      >
        ( 2 yrs )
      </span>
    </span>
    <span
      className="tag "
    >
      VCS Integration
      <span
        className="xp"
      >
        ( 1 yr )
      </span>
    </span>
    <span
      className="tag "
    >
      Nedbank iVeri Lite Integration
      <span
        className="xp"
      >
        ( 3 yrs )
      </span>
    </span>
    <span
      className="tag "
    >
      Mashreq Payment Gateway Integration
      <span
        className="xp"
      >
        ( 1 yr )
      </span>
    </span>
    <span
      className="tag "
    >
      Microsoft SQL Server
      <span
        className="xp"
      >
        ( 11 yrs )
      </span>
    </span>
  </div>
</div>
`;
